controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,623 @@
+			continue
+		}
+
+		m.snapshotCheckTaskQueue.Add(snapshotCheckTask{
+			volumeName:   engine.Spec.VolumeName,
+			snapshotName: snapshot.Name,
+			changeEvent:  false,
+		})
+	}
+}
+
+func (m *SnapshotMonitor) processNextWorkItem(id int) bool {
+	key, quit := m.snapshotCheckTaskQueue.Get()
+	if quit {
+		return false
+	}
+	defer m.snapshotCheckTaskQueue.Done(key)
+
+	task := key.(snapshotCheckTask)
+
+	dataIntegrity, err := m.getSnapshotDataIntegrity(task.volumeName)
+	if err != nil {
+		return true
+	}
+
+	if dataIntegrity == longhorn.SnapshotDataIntegrityDisabled {
+		return true
+	}
+
+	err = m.RunTask(task)
+	m.handleErr(err, key)
+
+	return true
+}
+
+func (m *SnapshotMonitor) handleErr(err error, key interface{}) {
+	if err == nil {
+		m.snapshotCheckTaskQueue.Forget(key)
+		return
+	}
+
+	if !strings.Contains(err.Error(), etypes.CannotRequestHashingSnapshotPrefix) {
+		m.snapshotCheckTaskQueue.Forget(key)
+		return
+	}
+
+	if m.snapshotCheckTaskQueue.NumRequeues(key) < snapshotHashMaxRetries {
+		logrus.Warnf("Error syncing snapshot check task %v: %v", key, err)
+		m.snapshotCheckTaskQueue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+
+	logrus.Warnf("Dropping hashing request of snapshot %v: %v", key, err)
+	m.snapshotCheckTaskQueue.Forget(key)
+}
+
+func (m *SnapshotMonitor) snapshotCheckWorker(id int) {
+	for m.processNextWorkItem(id) {
+	}
+}
+
+func (m *SnapshotMonitor) Close() {
+	m.logger.WithField("monitor", monitorName).Info("Closing snapshot monitor")
+
+	m.snapshotCheckTaskQueue.ShutDown()
+	m.checkScheduler.Stop()
+	m.quit()
+}
+
+func (m *SnapshotMonitor) Update(map[string]interface{}) error {
+	dataIntegrityCronJob, err := m.ds.GetSettingValueExisted(types.SettingNameSnapshotDataIntegrityCronJob)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get %v setting", types.SettingNameSnapshotDataIntegrityCronJob)
+	}
+
+	m.RLock()
+	if dataIntegrityCronJob == m.existingDataIntegrityCronJob && m.checkScheduler.Len() > 0 {
+		m.RUnlock()
+		return nil
+	}
+	m.RUnlock()
+
+	if m.checkScheduler.Len() > 0 {
+		m.checkScheduler.Remove(m.checkSnapshots)
+	}
+
+	job, err := m.checkScheduler.Cron(dataIntegrityCronJob).Do(m.checkSnapshots)
+	if err != nil {
+		return errors.Wrap(err, "failed to schedule snapshot check job")
+	}
+
+	m.Lock()
+	previousDataIntegrityCronJob := m.existingDataIntegrityCronJob
+	m.existingDataIntegrityCronJob = dataIntegrityCronJob
+	m.Unlock()
+
+	m.checkScheduler.StartAsync()
+
+	m.logger.WithField("monitor", monitorName).Infof("Cron is changed from %v to %v. Next snapshot check job will be executed at %v",
+		previousDataIntegrityCronJob, m.existingDataIntegrityCronJob, job.NextRun())
+
+	return nil
+}
+
+func (m *SnapshotMonitor) GetData() (interface{}, error) {
+	m.RLock()
+	defer m.RUnlock()
+	return m.SnapshotMonitorStatus, nil
+}
+
+func (m *SnapshotMonitor) addToInProgressSnapshotCheckTasks(snapshotName string) bool {
```

Comment:
Rename to `shouldAddToInProgressSnapshotCheckTasks`

Reply:
Updated. Thank you.